Add Serde support to HyperLogLog

mirosval opened the pull request:
I have a use-case where I need to send `HyperLogLog` over the network. In order to support that, I'd like to extend it with Serde support.

The PR:
* adds a new default `serde` feature
* adds a `hyperloglog::serde` module that depends on the feature and defines a new struct `HyperLogLogExport`, which just stores the `b` and the `registers` and implements Serde's `Serialize` and `Deserialize` traits. Additionally it adds a `from_exported_hyperloglog` method on `HyperLogLog` that is able to reconstruct self from export
* I don't know if there is a better way to tie the exporting and the reconstructing hashers, so right now it's up to the user to ensure the hasher is the same.
* Tests and docs
* Fixes various errors and clippy lints that popped up

Please let me know if this approach is acceptable and I can implement it also for some other structures, at a minimum, I also need it for the `BloomFilter`.

crepererum commented on src/hyperloglog.rs:
```diff
@@ -359,12 +362,102 @@ where
+    /// Captures the underlying HyperLogLog data such that it can be serialized and deserialized
+    /// using serde
+    #[derive(Clone, Serialize, Deserialize, Eq, PartialEq)]
+    pub struct HyperLogLogExport {
```
Instead of going through a helper struct, we should implement `Serialize` and `Deserialize` manually. In fact I think you don't need `serde-derive` here.

Regarding the hasher: either you need to constraint the `Serialize` impl to a hasher w/o any state or you also need to serialize the hasher. I think the latter one is technically the only proper way. That's important because the stdlib default hasher HAS state (the seed) and you shouldn't just drop that to the floor.

mirosval replied:
Hey, thanks for the review. I've implemented this in terms of Serde manually now, but I don't think this is the right approach. 

I see your point and it is a potential problem if the hashers diverge, but I would simply document it and leave it to the users to deal with. None of the hashers in the `std` or other crates I've seen typically implement Serde, so then you're stuck with the orphan rule. Many hashes are also often not stable across architectures, or even time, so it's generally not a good idea storing these.

With this new implementation I as a user can't opt out of serializing the hasher, wasting bandwidth, I also can't serialize the HLL to formats not supporting serde, like protobuf.

In my view it's much more preferable to either have this auxiliary struct, or simply just mark the `registers` and `b` as `pub` and call it a day.

crepererum commented on src/hyperloglog.rs:
```diff
@@ -152,15 +152,18 @@ where
+    /// Adds an already-hashed element to the HyperLogLog
+    ///
+    /// Note: Make sure to use the same hasher as the rest of the HyperLogLog when hashing values on your own
+    pub fn add_hashed(&mut self, hashed_value: u64) {
```
why is this public?

mirosval replied:
If I have a value that's already been hashed, there is no way for me to insert it otherwise. I can't unhash it..

crepererum commented on src/hyperloglog.rs:
```diff
@@ -720,4 +813,17 @@ mod tests {
+        let json = serde_json::to_string(&export).expect("can serialize to json");
+        let de_hll_export = serde_json::from_str(&json).expect("can deserialize from json");
+        let de_hll: HyperLogLog<str, _> =
+            HyperLogLog::from_exported_hyperloglog(de_hll_export, hasher);
```
so that's cheating: you bypassed the serialization of the hasher.

crepererum commented on src/hyperloglog.rs:
```diff
@@ -363,91 +364,200 @@ where
+        where
+            D: serde::Deserializer<'de>,
+        {
+            enum Field {
```
last nitpick: can you move the structs and impls into the `hyperloglog::serde` module? That makes the code a bit easier to read and the `serde` module prevents that stuff from leaking all over the place.

mirosval replied:
cool, I've moved the hll and data to a module and extracted serde. I've added a public getter for the registers, because as I said, having just a direct serde impl doesn't actually help for my use case, which is kind of why I started the PR in the first place.

mirosval commented on src/hyperloglog/serde.rs:
fixed

mirosval commented:
> clippy is failing

Fixed. I've also realized that I'd forgotten to include a way to reconstruct the instance from the extracted registers, so I've added that

crepererum commented on src/hyperloglog/mod.rs:
```diff
@@ -114,14 +120,27 @@ where
+    }
+
+    /// Same as `new` but with pre-initialized registers and a specific `BuildHasher`.
+    pub fn with_registers_and_hash(b: usize, registers: Vec<u8>, buildhasher: B) -> Self {
```
Why is this required? As far as I can see, the `serde` impl doesn't need that. Also there's no way to actually read the registers, so this seems rather useless.

mirosval replied:
I've added the corresponding `registers()` [here](https://github.com/crepererum/pdatastructs.rs/pull/129/files#diff-37372a8fbd5759f7e6fef0e85f6ffa4783c80319a4a6a3806009270dbfa63578R165-R167). 

As I mentioned in the other replies to your comments, what I actually need to do is to serialize this structure using Protobuf. The initial design was to introduce an auxiliary structure that would decouple the actual HLL parameters (`b` and `registers`) from the `PhantomData` and the `T` as those are only needed for hashing and not actually relevant to the HLL (if we view the HLL as if it was working with already hashed values).

However the requirement to implement `serde` for the `HyperLogLog` structure directly leaves me without the ability to do the protobuf serialization. So I introduced a method to both retrieve the `registers` and `b`, as well as a new constructor that takes these 2 params along with a hasher.

crepererum replied:
Alright, sounds good to me then.

crepererum reviewed (approved):
Thank you very much, esp. for leaving the code in a better shape than you've found it :heart:

crepererum commented:
bors r+
